pkg/policy/k8s/service.go

```diff
@@ -0,0 +1,278 @@
+			cnp.ObjectMeta.Namespace,
+			cnp.ObjectMeta.Name,
+		)
+
+		errs = append(errs, p.resolveCiliumNetworkPolicyRefs(cnp, key, initialRecvTime, resourceID))
+	}
+	return errors.Join(errs...)
+}
+
+// resolveToServices translates all ToServices rules found in the provided CNP
+// and to corresponding ToCIDRSet rules. Mutates the passed in cnp in place.
+func (p *PolicyWatcher) resolveToServices(cnp *types.SlimCNP) {
+	// The svcEndpoints internally associates each ToService selector with
+	// the list of matching service endpoints.
+	svcEndpoints := emptyMatchedEndpoints()
+
+	// We consult the service cache to obtain the service endpoints
+	// which are selected by the ToServices selectors found in the CNP.
+	// The selected endpoints are stored in svcEndpoints.
+	p.svcCache.ForEachService(func(svcID k8s.ServiceID, svc *k8s.Service, eps *k8s.Endpoints) bool {
+		if p.isSelectableService(svc) {
+			svcEndpoints.extractMatches(cnp.Spec, svcID, svc, eps)
+			for _, spec := range cnp.Specs {
+				svcEndpoints.extractMatches(spec, svcID, svc, eps)
+			}
+		}
+
+		return true
+	})
+
+	// Once we have collected all service endpoints  we are translating each
+	// rule by producing a new ToCIDRSet entry that corresponds to the
+	// rule's ToServices entry. This mutates cnp in place.
+	svcEndpoints.translateRule(cnp.Spec)
+	for _, spec := range cnp.Specs {
+		svcEndpoints.translateRule(spec)
+	}
+}
+
+// cnpMatchesService returns true if the cnp contains a ToServices rule which
+// matches the provided service svcID/svc
+func (p *PolicyWatcher) cnpMatchesService(cnp *types.SlimCNP, svcID k8s.ServiceID, svc *k8s.Service) bool {
+	if !p.isSelectableService(svc) {
+		return false
+	}
+
+	if hasMatchingToServices(cnp.Spec, svcID, svc) {
+		return true
+	}
+
+	for _, spec := range cnp.Specs {
+		if hasMatchingToServices(spec, svcID, svc) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// specHasMatchingToServices returns true if the rule contains a ToServices rule which
+// matches the provided service svcID/svc
+func hasMatchingToServices(spec *api.Rule, svcID k8s.ServiceID, svc *k8s.Service) bool {
+	if spec == nil {
+		return false
+	}
+	for _, egress := range spec.Egress {
+		for _, toService := range egress.ToServices {
+			if sel := toService.K8sServiceSelector; sel != nil {
+				if serviceSelectorMatches(sel, svcID, svc) {
+					return true
+				}
+			} else if ref := toService.K8sService; ref != nil {
+				if serviceRefMatches(ref, svcID) {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// serviceSelectorMatches returns true if the ToServices k8sServiceSelector
+// matches the labels of the provided service svc
+func serviceSelectorMatches(sel *api.K8sServiceSelectorNamespace, svcID k8s.ServiceID, svc *k8s.Service) bool {
+	es := api.EndpointSelector(sel.Selector)
+	es.SyncRequirementsWithLabelSelector()
+	esMatches := es.Matches(labels.Set(svc.Labels))
+	return esMatches &&
+		(sel.Namespace == svcID.Namespace || sel.Namespace == "")
```

Comment:
minor nit: can save a lot of comparisons by bailing early if namespace doesn't match.